### PR TITLE
Java: Add client configuration for subscribing to channels.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Ref: https://github.com/aws/glide-for-redis/pull/1602/ and https://github.com/aws/glide-for-redis/pull/1616 and https://github.com/aws/glide-for-redis/pull/1650

Tested manually with
```java
    @Test
    @SneakyThrows
    public void basic_client() {
        BiConsumer<Object, Object> callback =
            (msg, context) -> System.out.printf("Received %s, context %s\n", msg, context);

        var regularClient =
                RedisClient.CreateClient(
                                RedisClientConfiguration.builder()
                                        .address(NodeAddress.builder().port(6379).build())
                                        .requestTimeout(3000)
                                        .subscriptionConfiguration(
                                                StandaloneSubscriptionConfiguration.builder()
                                                        .subscription(EXACT, "ch1")
                                                        .subscription(PATTERN, "chat*")
                                                        .callback(callback)
                                                        .build())
                                        .build())
                        .get();

        Thread.sleep(100500);
    }
```

And with `redis-cli`: `publish ch1 $message`

TODO (in scope of this PR):
- [x] Complete docs

Next feature parts (other PRs):
- [ ] Handling received pushes - #385
- [ ] Sending messages - #391
- [ ] Tests
- [ ] Examples

### CODE REVIEW

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 1]
```diff
@@ -244,6 +248,10 @@ public abstract class BaseClient
     protected final ConnectionManager connectionManager;
     protected final CommandManager commandManager;
 
+    @Setter(AccessLevel.PROTECTED)
+    @Accessors(chain = true)
+    protected BaseSubscriptionConfiguration<?> subscriptionConfiguration;
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:253]: Probably, this part will be reworked in the follow-up PR.

File: java/client/src/main/java/glide/api/models/configuration/RedisClientConfiguration.java [hunk 2]
```diff
@@ -13,4 +18,36 @@ public class RedisClientConfiguration extends BaseClientConfiguration {
...
+    public abstract static class RedisClientConfigurationBuilder<
+                    C extends RedisClientConfiguration, B extends RedisClientConfigurationBuilder<C, B>>
+            extends BaseClientConfigurationBuilder<C, B> {
+        /**
+         * Add a subscription to a channel or to multiple channels if {@link
+         * PubSubStandaloneChannelMode#PATTERN} is used.
+         */
+        public B addSubscription(PubSubStandaloneChannelMode mode, String channelOrPattern) {
+            if (subscriptions == null) {
+                subscriptions = new HashMap<>(2);
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/configuration/RedisClientConfiguration.java]: I reworked this.
There are only 2 channel types.

File: java/client/src/main/java/glide/connectors/handlers/CallbackDispatcher.java [hunk 3]
```diff
@@ -78,6 +80,13 @@ public void completeRequest(Response response) {
             distributeClosingException(response.getClosingError());
             return;
         }
+        // ignore pushes to avoid crashed when received
+        // TODO rework
+        if (response.getIsPush()) {
+            var data = valueFromPointer(response.getRespPointer());
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/connectors/handlers/CallbackDispatcher.java:86]: Yes, I added this to trigger this line https://github.com/Bit-Quill/glide-for-redis/pull/381/files#diff-5973d8064e5f7884b62606c357768d021fac02a34c589f3ce187e949f36deab2R99

File: java/client/src/main/java/glide/api/RedisClient.java [hunk 4]
```diff
@@ -75,7 +75,11 @@ protected RedisClient(ConnectionManager connectionManager, CommandManager comman
      */
     public static CompletableFuture<RedisClient> CreateClient(
             @NonNull RedisClientConfiguration config) {
-        return CreateClient(config, RedisClient::new);
+        return CreateClient(config, RedisClient::new)
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/RedisClient.java:78]: Yes and no
Part of the config goes to GLIDE core inside protobuf ConnectionRequest, but the callback should be stored inside the client.

**Review by aaron-congo — APPROVED** [review]

File: java/client/src/main/java/glide/managers/ConnectionManager.java [hunk 5]
```diff
@@ -141,6 +144,19 @@ private ConnectionRequest.Builder setupConnectionRequestBuilderRedisClient(
             connectionRequestBuilder.setDatabaseId(configuration.getDatabaseId());
         }
 
+        if (configuration.getSubscriptionConfiguration() != null) {
+            var subscriptionsBuilder = PubSubSubscriptions.newBuilder();
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/managers/ConnectionManager.java:148]: I would move this logic to the client, and create & the subscriptions in the constructor.  Then you can just use a getter here to grab the subscriptions.  

Would reduce duplicate code below.

File: java/client/src/main/java/glide/managers/ConnectionManager.java [hunk 6]
```diff
@@ -149,13 +165,25 @@ private ConnectionRequest.Builder setupConnectionRequestBuilderRedisClient(
...
-    private ConnectionRequestOuterClass.ConnectionRequest.Builder
-            setupConnectionRequestBuilderRedisClusterClient(
-                    RedisClusterClientConfiguration configuration) {
+    private ConnectionRequest.Builder setupConnectionRequestBuilderRedisClusterClient(
+            RedisClusterClientConfiguration configuration) {
         ConnectionRequest.Builder connectionRequestBuilder =
                 setupConnectionRequestBuilderBaseConfiguration(configuration);
         connectionRequestBuilder.setClusterModeEnabled(true);
 
+        if (configuration.getSubscriptionConfiguration() != null) {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/managers/ConnectionManager.java:174]: seems like duplicate code

**Review by acarbonetto — APPROVED** [review]

File: java/client/src/main/java/glide/api/models/configuration/BaseSubscriptionConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+     * </ol>
+     */
+    // TODO change message type
+    public interface MessageCallback extends BiConsumer<Object, Object> {}
+
+    /**
+     * Optional callback to accept the incoming messages. See {@link MessageCallback}.<br>
+     * If not set, messages will be available via TODO method.<br>
+     */
+    protected final Optional<MessageCallback> callback;
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/configuration/BaseSubscriptionConfiguration.java:43]: feels like callback subscription configuration and queued configuration could be separate objects

File: java/client/src/main/java/glide/managers/ConnectionManager.java (hunk 5 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/managers/ConnectionManager.java:148]: `PubSubSubscriptions` is a protobuf class. It is a right place for that.

File: java/client/src/main/java/glide/managers/ConnectionManager.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/managers/ConnectionManager.java:174]: Unfortunately I didn't find a wayt to move it to another function. Classes are different for standalone and cluster clients.

File: java/client/src/main/java/glide/api/models/configuration/BaseSubscriptionConfiguration.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/configuration/BaseSubscriptionConfiguration.java:43]: How do you see that?
Is it good enough?
```java
        var regularClient =
                RedisClient.CreateClient(
                                RedisClientConfiguration.builder()
                                        .address(NodeAddress.builder().port(6379).build())
                                        .requestTimeout(3000)
                                        .subscriptionConfiguration(
                                                StandaloneSubscriptionConfiguration.builder()
                                                        .subscription(PubSubChannelMode.EXACT, "ch1")
                                                        .subscription(PubSubChannelMode.PATTERN, "chat*")
                                                                        //   from there
                                                        .build())
                                        .subscriptionCallback(callback) // I moved this
                                        .build())
                        .get();
```

**Pull request merged**